client/network/src/discovery.rs

```diff
@@ -629,20 +629,33 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 			FromSwarm::ListenerError(e) => {
 				self.kademlia.on_swarm_event(FromSwarm::ListenerError(e));
 			},
-			FromSwarm::ExpiredExternalAddr(e) => {
+			FromSwarm::ExternalAddrExpired(e) => {
 				// We intentionally don't remove the element from `known_external_addresses` in
 				// order to not print the log line again.
 
-				self.kademlia.on_swarm_event(FromSwarm::ExpiredExternalAddr(e));
+				self.kademlia.on_swarm_event(FromSwarm::ExternalAddrExpired(e));
 			},
 			FromSwarm::NewListener(e) => {
 				self.kademlia.on_swarm_event(FromSwarm::NewListener(e));
 			},
 			FromSwarm::ExpiredListenAddr(e) => {
 				self.kademlia.on_swarm_event(FromSwarm::ExpiredListenAddr(e));
 			},
-			FromSwarm::NewExternalAddr(e @ NewExternalAddr { addr }) => {
-				let new_addr = addr.clone().with(Protocol::P2p(self.local_peer_id.into()));
+			FromSwarm::NewExternalAddrCandidate(e) => {
+				self.kademlia.on_swarm_event(FromSwarm::NewExternalAddrCandidate(e));
+			},
+			FromSwarm::AddressChange(e) => {
+				self.kademlia.on_swarm_event(FromSwarm::AddressChange(e));
+			},
+			FromSwarm::NewListenAddr(e) => {
+				self.kademlia.on_swarm_event(FromSwarm::NewListenAddr(e));
+
+				if let Some(ref mut mdns) = self.mdns {
+					mdns.on_swarm_event(FromSwarm::NewListenAddr(e));
```

Comment:
We can use [libp2p_swarm](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/index.html)::[behaviour](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/behaviour/index.html)::[toggle](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/behaviour/toggle/index.html)::[Toggle](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/behaviour/toggle/struct.Toggle.html#) instead of `Option<TokioMdns>`. It will remove ifs because they are hidden inside `Toggle::on_*` methods.

Reply:
Good idea 👍 Will do in a separate PR